cmd/registry/scoring/score.go

```diff
@@ -126,121 +132,164 @@ func matchResourceWithTarget(targetPattern *rpc.ResourcePattern, resourceName pa
 	client connection.Client,
 	formula *rpc.ScoreFormula,
-	resource patterns.ResourceInstance) (interface{}, error) {
+	resource patterns.ResourceInstance,
+	scoreArtifact *rpc.Artifact,
+	takeAction bool) (interface{}, bool, error) {
 	extendedArtifact, err := patterns.SubstituteReferenceEntity(formula.GetArtifact().GetPattern(), resource.ResourceName())
 	if err != nil {
-		return nil, fmt.Errorf("invalid score_formula.artifact.pattern: %s for {%v}, %s", formula.GetArtifact().GetPattern(), formula, err)
-	}
-	artifactName, err := names.ParseArtifact(extendedArtifact.String())
-	if err != nil {
-		return nil, fmt.Errorf("invalid score_formula.artifact.pattern: %s for {%v}, %s", formula.GetArtifact().GetPattern(), formula, err)
+		return nil, false, fmt.Errorf("invalid score_formula.artifact.pattern: %s for {%v}, %s", formula.GetArtifact().GetPattern(), formula, err)
 	}
 	if formula.GetScoreExpression() == "" {
-		return nil, fmt.Errorf("missing score_formula.score_expression for {%v}", formula)
+		return nil, false, fmt.Errorf("missing score_formula.score_expression for {%v}", formula)
 	}
 
 	// Fetch the artifact
-	var contents []byte
-	var mimeType string
-	err = core.GetArtifact(ctx, client, artifactName, true, func(artifact *rpc.Artifact) error {
-		contents = artifact.GetContents()
-		mimeType = artifact.GetMimeType()
-		return nil
-	})
+	artifact, err := getArtifact(ctx, client, extendedArtifact.String(), true)
 	if err != nil {
-		return nil, fmt.Errorf("failed to fetch artifact %s: %s", artifactName, err)
+		return nil, false, fmt.Errorf("failed to fetch artifact %s: %s", extendedArtifact.String(), err)
 	}
 
-	// TODO: Add timestamp check. Compute the score only if the artifact has an update since the last score calculation
+	// Update required tells the calling function if the score artifact needs to be updated
+	updateRequired := takeAction || scoreArtifact.GetUpdateTime().AsTime().Before(artifact.GetUpdateTime().AsTime())
+
+	// Apply the scoreExpression by default. This value will be required by the rollup_formula in the case where
+	// another formula from rollup_formula.score_formulas makes the score outdated.
 
 	// Convert artifact contents to map[string]interface{}
-	artifactMap, err := getMap(contents, mimeType)
+	artifactMap, err := getMap(artifact.GetContents(), artifact.GetMimeType())
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	// Apply the score_expression
-	return evaluateScoreExpression(formula.GetScoreExpression(), artifactMap)
+	value, err := evaluateScoreExpression(formula.GetScoreExpression(), artifactMap)
+	if err != nil {
+		return nil, false, err
+	}
+	return value, updateRequired, nil
 }
 
 func processRollUpFormula(
 	ctx context.Context,
 	client connection.Client,
 	formula *rpc.RollUpFormula,
-	resource patterns.ResourceInstance) (interface{}, error) {
+	resource patterns.ResourceInstance,
+	scoreArtifact *rpc.Artifact,
+	takeAction bool) (interface{}, bool, error) {
```

Comment:
It's unclear what this boolean is supposed to show. Can you add a struct so these values can have a name and documentation associated with them? This will be valuable for the `interface{}` too.

Reply:
Done, hope this helps with understanding the code.